pyvortex/src/dataset.rs

```diff
@@ -3,69 +3,57 @@ use std::sync::Arc;
 use arrow::array::RecordBatchReader;
 use arrow::datatypes::SchemaRef;
 use arrow::pyarrow::{IntoPyArrow, ToPyArrow};
+use futures::TryStreamExt;
 use pyo3::exceptions::PyTypeError;
 use pyo3::prelude::*;
 use pyo3::types::PyString;
+use vortex::array::ChunkedArray;
 use vortex::arrow::infer_schema;
-use vortex::dtype::{DType, FieldName};
+use vortex::buffer::Buffer;
+use vortex::dtype::FieldName;
 use vortex::error::VortexResult;
-use vortex::expr::RowFilter;
-use vortex::file::{
-    read_initial_bytes, LayoutContext, LayoutDeserializer, Projection, VortexReadArrayStream,
-    VortexReadBuilder, VortexRecordBatchReader,
-};
+use vortex::expr::{get_item, ident, pack, ExprRef, Identity};
+use vortex::file::v2::io::file::FileIoDriver;
+use vortex::file::v2::{Scan, VortexFile, VortexOpenOptions};
+use vortex::file::VortexRecordBatchReader;
 use vortex::io::{ObjectStoreReadAt, TokioFile, VortexReadAt};
 use vortex::sampling_compressor::ALL_ENCODINGS_CONTEXT;
-use vortex::ArrayData;
+use vortex::stream::ArrayStream;
+use vortex::{ArrayData, IntoArrayData, IntoArrayVariant};
 
 use crate::expr::PyExpr;
 use crate::object_store_urls::vortex_read_at_from_url;
 use crate::{PyArray, TOKIO_RUNTIME};
 
-pub async fn layout_stream_from_reader<T: VortexReadAt + Unpin>(
-    reader: T,
-    projection: Projection,
-    row_filter: Option<RowFilter>,
+pub async fn read_array_from_reader<T: VortexReadAt + Unpin + 'static>(
+    vortex_file: &VortexFile<FileIoDriver<T>>,
+    projection: ExprRef,
+    filter: Option<ExprRef>,
     indices: Option<ArrayData>,
-) -> VortexResult<VortexReadArrayStream<T>> {
-    let mut builder = VortexReadBuilder::new(
-        reader,
-        LayoutDeserializer::new(
-            ALL_ENCODINGS_CONTEXT.clone(),
-            LayoutContext::default().into(),
-        ),
-    )
-    .with_projection(projection);
-
-    if let Some(row_filter) = row_filter {
-        builder = builder.with_row_filter(row_filter);
+) -> VortexResult<ArrayData> {
+    let mut scan = Scan::new(projection);
+
+    if let Some(filter) = filter {
+        scan = scan.with_filter(filter);
     }
 
     if let Some(indices) = indices {
-        builder = builder.with_indices(indices);
+        let indices = Buffer::from_iter(indices.into_primitive()?.as_slice::<u64>().to_vec());
```

Comment:
I think the preferred way to say this is `indices.into_primitive()?.into_buffer()`. If indices was compressed and `into_primitive` decompressed it, then you take ownership of that decompressed buffer without a copy. The `as_slice.to_vec` makes a copy even though you might be the unique owner of the primitive array's buffer.

Reply:
100%